### DISCUSSION TITLE
[nginx-proxy-manager] Cannot generate Let's Encrypt certification.

### DISCUSSION THREAD

**Discussion by hilee0602** [tteck/Proxmox#1196] [Help Needed]
Hi.

Recently I reinstalled my server and reinstalled npm.
And I try to generate new Let's Encrypt certification on npm, then I saw below error message.

"
Error: Command failed: certbot certonly --config "/etc/letsencrypt.ini" --cert-name "npm-2" --agree-tos --email "[My Email]" --domains "[Domain]" --authenticator dns-cloudflare --dns-cloudflare-credentials "/etc/letsencrypt/credentials/credentials-2"
usage: 
  certbot [SUBCOMMAND] [options] [-d DOMAIN] [-d DOMAIN] ...

Certbot can obtain and install HTTPS/TLS/SSL certificates.  By default,
it will attempt to use a webserver both for obtaining and installing the
certificate. 
certbot: error: unrecognized arguments: --dns-cloudflare-credentials /etc/letsencrypt/credentials/credentials-2

    at ChildProcess.exithandler (node:child_process:402:12)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1100:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5)
"

Could you help me how can I fix this problem??

Thanks.

**Comment by tteck** [comment]
[c43dcea](https://github.com/tteck/Proxmox/commit/c43dcea48ed2e9e45a5a698bac954b8258e995fd) should fix your issue. If not, let me know.

**Reply by hilee0602** [reply]
Thank you for reply.

Now I'm try to reinstall npm.

**Reply by tteck** [reply]
One option would have been to execute this command:
```
python3 -m pip uninstall certbot-dns-cloudflare
```
However, a fresh installation should also do the trick.

**Reply by hilee0602** [reply]
Hi. tteck.

Unfortunately it was not fixed by [c43dcea](https://github.com/tteck/Proxmox/commit/c43dcea48ed2e9e45a5a698bac954b8258e995fd) commit. 
So I installed "certbot-dns-cloudflare" like below steps and then I can make certificate file.

1. Install npm using "bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/nginxproxymanager-v5.sh)"" command.
2. Access npm terminal on proxmox web shell.
3. Type "python3 -m pip install --no-cache-dir cffi certbot certbot-dns-cloudflare" command on terminal.
4. Make Let's Encrypt certification files on npm web.
![image](https://user-images.githubusercontent.com/72373977/226216938-3b4a39af-a2ed-4893-adeb-51de8a0c11ba.png)


So I think something happened when install "certbot-dns-cloudflare" in shell script.

Thank you.

**Reply by tteck** [reply]
I'm a bit uncertain, do you intend to utilize the Cloudflare DNS challenge?

**Reply by tteck** [reply]
If not, you can skip the second and third steps.

**Reply by hilee0602** [reply]
I checked "Use a DNS Challenge" option when I make certificate file.
![image](https://user-images.githubusercontent.com/72373977/226218635-8be39301-7178-4d33-ab44-5052557dab49.png)

**Reply by Vildmaskine** [reply]
Thanks, this fixed it for me.

**Reply by rcarteraz** [reply]
This also fixed it for me. Thank you!

**Comment by rayd123** [comment]
Hi tteck,

Firstly, thank you for sharing with the community and your obvious hard work and commitment to this project. 

I had the same issue --"error: unrecognized arguments: --dns-namecheap-credentials /etc/letsencrypt/credentials/credentials-1"

I tried what was mentioned in #1282 using `nano /opt/certbot/pyvenv.cfg` and it did not work. 

I love it when stuff does not work as that provides the maximum learning curve behind the scenes. 

I tired reinstalling as I moved this from one proxmox node to another using --: `bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/nginxproxymanager.sh)"`

In the end the issue was resolved by --: `python3 -m pip install --no-cache-dir cffi certbot certbot-dns-namecheap `

Thanks again!!!

For completeness...
```
Requirement already satisfied: cffi in /usr/lib/python3/dist-packages (1.15.1)
Requirement already satisfied: certbot in /usr/lib/python3/dist-packages (2.1.0)
Collecting certbot-dns-namecheap
  Downloading certbot-dns-namecheap-1.0.0.tar.gz (9.4 kB)
  Preparing metadata (setup.py) ... done
Collecting PyNamecheap
  Downloading PyNamecheap-0.0.3.tar.gz (5.7 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: acme>=0.21.1 in /usr/lib/python3/dist-packages (from certbot-dns-namecheap) (2.1.0)
Collecting dns-lexicon>=2.2.1
  Downloading dns_lexicon-3.17.0-py3-none-any.whl (243 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.9/243.9 kB 10.7 MB/s eta 0:00:00
Collecting mock
  Downloading mock-5.1.0-py3-none-any.whl (30 kB)
Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (from certbot-dns-namecheap) (66.1.1)
Collecting zope.interface
  Downloading zope.interface-6.1-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (249 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 249.1/249.1 kB 18.6 MB/s eta 0:00:00
Requirement already satisfied: beautifulsoup4>=4 in /usr/lib/python3/dist-packages (from dns-lexicon>=2.2.1->certbot-dns-namecheap) (4.11.2)
Requirement already satisfied: cryptography>=2 in /usr/lib/python3/dist-packages (from dns-lexicon>=2.2.1->certbot-dns-namecheap) (38.0.4)
Requirement already satisfied: dnspython>=2 in /usr/local/lib/python3.11/dist-packages (from dns-lexicon>=2.2.1->certbot-dns-namecheap) (2.4.2)
Collecting pyotp>=2
  Downloading pyotp-2.9.0-py3-none-any.whl (13 kB)
Requirement already satisfied: pyyaml>=3 in /usr/lib/python3/dist-packages (from dns-lexicon>=2.2.1->certbot-dns-namecheap) (6.0)
Requirement already satisfied: requests>=2 in /usr/lib/python3/dist-packages (from dns-lexicon>=2.2.1->certbot-dns-namecheap) (2.28.1)
Collecting tldextract>=2
  Downloading tldextract-5.1.1-py3-none-any.whl (97 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.7/97.7 kB 39.2 MB/s eta 0:00:00
Requirement already satisfied: idna in /usr/lib/python3/dist-packages (from tldextract>=2->dns-lexicon>=2.2.1->certbot-dns-namecheap) (3.3)
Collecting requests-file>=1.4
  Downloading requests_file-1.5.1-py2.py3-none-any.whl (3.7 kB)
Collecting filelock>=3.0.8
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Requirement already satisfied: six in /usr/lib/python3/dist-packages (from requests-file>=1.4->tldextract>=2->dns-lexicon>=2.2.1->certbot-dns-namecheap) (1.16.0)
Building wheels for collected packages: certbot-dns-namecheap, PyNamecheap
  Building wheel for certbot-dns-namecheap (setup.py) ... done
  Created wheel for certbot-dns-namecheap: filename=certbot_dns_namecheap-1.0.0-py2.py3-none-any.whl size=9453 sha256=144b77e9732ea66b12bba5c9bb08a40454ff35377b6923e085a02293f566304e
  Stored in directory: /tmp/pip-ephem-wheel-cache-7bkjyw6u/wheels/a3/a9/55/b167ce2b29a5ff39939b2c92dfa0129fff8df86ec0d1b23df8
  Building wheel for PyNamecheap (setup.py) ... done
  Created wheel for PyNamecheap: filename=PyNamecheap-0.0.3-py3-none-any.whl size=6100 sha256=1a90dffff46c9d38f989c298e89fef7041fe260449ce76c1a2e04bcf3307790a
  Stored in directory: /tmp/pip-ephem-wheel-cache-7bkjyw6u/wheels/a3/94/b0/9b76b5661ea896b49428475d5e50b378813a0f787304e7096f
Successfully built certbot-dns-namecheap PyNamecheap
Installing collected packages: zope.interface, requests-file, pyotp, PyNamecheap, mock, filelock, tldextract, dns-lexicon, certbot-dns-namecheap
Successfully installed PyNamecheap-0.0.3 certbot-dns-namecheap-1.0.0 dns-lexicon-3.17.0 filelock-3.13.1 mock-5.1.0 pyotp-2.9.0 requests-file-1.5.1 tldextract-5.1.1 zope.interface-6.1
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@nginxproxymanager:~#
```
